### PR TITLE
Add reload gateway command to the tykctl

### PR DESCRIPTION
**Pull request by yurisasuke**
## Description
- This pull request adds the reload command to the gateway service. Users can now reload their gateway with the tykctl.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [X] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork,
  don't come from your `master`!
- [X] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you
  started *your change* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
    - [ ] I have manually updated the README(s)/documentation accordingly.
    - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
    - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
    - [ ] `go fmt -s`
    - [ ] `go vet`

### CODE REVIEW

File: gatewaycmd/gateway.go [hunk 1]
```diff
@@ -1,12 +1,15 @@
 package gatewaycmd
 
 import (
+	"github.com/TykTechnologies/tykctl/gatewaycmd/reload"
```
Comment (mitjaziv) [inline gatewaycmd/gateway.go:6]: I would suggest to put Tyk import in separate group after 3rd party imports.

File: .github/workflows/generate.yml [hunk 2]
```diff
@@ -7,7 +7,7 @@ jobs:
   test:
     runs-on: ubuntu-20.04
     env:
-      GOPRIVATE: github.com/TykTechnologies/cloud-sdk
+      GOPRIVATE: github.com/TykTechnologies/cloud-sdk,github.com/TykTechnologies/gateway-sdk
```
Comment (mitjaziv) [inline .github/workflows/generate.yml]: You can do just `GOPRIVATE: github.com/TykTechnologies`

File: gatewaycmd/mock.go [hunk 3]
```diff
@@ -0,0 +1,3 @@
+package gatewaycmd
+
+//go:generate mockgen -package=mock -destination=./mocks/gateway_mock.go github.com/TykTechnologies/gateway-sdk/pkg/apim HotReloadAPI
```
Comment (mitjaziv) [inline gatewaycmd/mock.go:1]: I would suggest to have //go:generate on the first line of file, but that is just personal preference.

File: gatewaycmd/reload/reload.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
+package reload
+
+import (
+	"context"
+	"errors"
+	"github.com/TykTechnologies/gateway-sdk/pkg/apim"
```
Comment (mitjaziv) [inline gatewaycmd/reload/reload.go:11]: I would suggest to have three separete groups for imports:

```
import (
        "context"
	"errors"
	"net/http"
	
	"github.com/spf13/cobra"
	"github.com/spf13/pflag"
	
	"github.com/TykTechnologies/gateway-sdk/pkg/apim"
	"github.com/TykTechnologies/tykctl/gatewaycmd/shared"
	"github.com/TykTechnologies/tykctl/internal"
)
```

File: internal/apim_factory.go [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package internal
+
+import "github.com/TykTechnologies/gateway-sdk/pkg/apim"
+
+type ApimFactory struct {
```
Comment (mitjaziv) [inline internal/apim_factory.go]: I would suggest renaming `ApimFactory` to `ApimClient`. 

Factory is specific method pattern approach, which is not used here, as a result it makes code confusing.

File: gatewaycmd/shared/shared.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"github.com/TykTechnologies/gateway-sdk/pkg/apim"
+	"github.com/spf13/viper"
+)
+
+var (
+	gatewaySecret     = "gateway.secret"
+	gatewayUrls       = "gateway.urls"
+	xTykAuthorization = "x-tyk-authorization"
```
Comment (mitjaziv) [inline gatewaycmd/shared/shared.go]: Would suggest to change `xTykAuthorization` to constant.

**Review by mitjaziv — CHANGES_REQUESTED** [review]
Please add `golangci-linter` to repository as there are syntax issue, which would be reported by linter.

I would suggest to use some more new lines between unrelated blocks of code, to make code more readable.

All in all approach looks good. 🚀

**Comment by yurisasuke** [comment]
> Please add `golangci-linter` to repository as there are syntax issue, which would be reported by linter.
> 
> I would suggest to use some more new lines between unrelated blocks of code, to make code more readable.
> 
> All in all approach looks good. 🚀

Thank you for the review. I have added a more strict golangci-linter configuration to catch the syntax issues. I have also added [Whitespace Linter ](https://github.com/bombsimon/wsl) to help with new lines between unrelated blocks of code.

File: gatewaycmd/reload/reload.go (hunk 4 above)
Comment (yurisasuke) [reply gatewaycmd/reload/reload.go:11]: I have added    
`  goimports:
    local-prefixes: "github.com/TykTechnologies"` in the  .golangci.yml to help with this.
:https://github.com/TykTechnologies/tykctl/blob/e6e1846b29dad7760ce248677db3dc3391f08ed3/.golangci.yml#L103

File: cloudcmd/rbac.go [hunk 7]
```diff
@@ -20,9 +22,9 @@ type CloudRbac struct {
 	Config         internal.UserConfig
 }
 
-func NewCloudRbac(MinAllowedUser Permission, config internal.UserConfig) CloudRbac {
+func NewCloudRbac(minAllowedUser Permission, config internal.UserConfig) CloudRbac {
```
Comment (mitjaziv) [inline cloudcmd/rbac.go:25]: Nit: Minor comment, RBAC is acronym like API,ID... (Role Based Access Control), as I saw that you changed to upper case for acronyms.

**Review by mitjaziv — APPROVED** [review]

**Pull request merged**